Transaction API's for Creating, Updating, and Deleting Transactions

Hamza-Chaudhary65 opened the pull request:
`Creating Transactions:`  The create_transaction method creates a new transaction record in the database.

`Updating Balances on Creation:`  The update_balance_on_transaction_creation method updates the account balance based on the transaction category (withdrawal or deposit) when a new transaction is created.

`Handling New Transactions:`  The handle_new_transaction method combines creating a transaction and updating the  account balance.

`Delete Transactions:`  The delete_transaction method deletes a transaction and adjusts the account balance accordingly.

`Updating Transactions:`  The update_transaction method handles  updates to transactions, including:

Updating description
- Date
- Amount
- Category 
- Account.

Update account balances if the account, category, or amount changes.

Separate methods 

- update_balance_if_account_changed
-  update_balance_if_category_changed
- update_balance_if_amount_changed  
 handle specific balance updates based on changes.

Hamza-Chaudhary65 committed:
created helper in bankaccounts. any changes in bank accounts would me managed by bank_account helper

Hamza-Chaudhary65 committed:
views

Hamza-Chaudhary65 committed:
Views api for transactions.
updated transaction helper with implemented logic

Hamza-Chaudhary65 committed:
Views api for Transactions

Hamza-Chaudhary65 committed:
eof

sheralim012 commented on fintrack/accounts/account_helper.py:
```diff
@@ -0,0 +1,10 @@
+
+
+class AccountHelper():
```
is parenthesis necessary here?

sheralim012 commented on fintrack/accounts/management/commands/populate_data.py:
```diff
@@ -0,0 +1,35 @@
+class Command(BaseCommand):
+
+    def handle(self, *args, **kwargs):     
+        # user, _ = User.objects.get_or_create(username='hamza65', password='hamza')
```
why these are commented out?

Hamza-Chaudhary65 replied:
I was using this just for testing purpose

sheralim012 commented on fintrack/accounts/models.py:
```diff
@@ -0,0 +1,24 @@
+class Account(models.Model):
+   
+
+    holder_name = models.CharField(max_length=100)
+    acc_number = models.CharField(max_length=20, primary_key=True)
```
I want you to read about banks terminologies and use those here like title, number etc

sheralim012 commented on fintrack/choices.py:
```diff
@@ -0,0 +1,12 @@
+
+
+class TransactionCategoriesChoices(models.TextChoices):
+    Withdrawl = "Withdrawl", _("Withdrawl")
```
```suggestion
    withdraw = "WITHDRAW", "Withdraw"
```

sheralim012 replied:
are we supporting multiple languages or doing translation in other places if not then remove gettext support

sheralim012 commented on fintrack/transactions/models.py:
```diff
@@ -0,0 +1,23 @@
+
+class Transaction(models.Model):
+
+    description = models.CharField(max_length=300, blank=True)
```
description should be text field instead

sheralim012 commented on fintrack/transactions/transaction_helper.py:
```diff
@@ -0,0 +1,134 @@
+        )  
+    
+    def update_balance_on_transaction_creation(self, account, transaction_updated_data):
+        if transaction_updated_data["category"] == WITHDRAWAL_CATEGORY:
```
read about database transactions

sheralim012 commented on fintrack/transactions/views.py:
```diff
@@ -0,0 +1,51 @@
+
+
+@api_view(['GET', 'PUT', 'DELETE'])
+def transaction_detail(request, pk):
```
it would be better to create separate views so they are doing specific things

sheralim012 commented on fintrack/transactions/transaction_helper.py:
```diff
@@ -0,0 +1,134 @@
+WITHDRAWAL_CATEGORY = "withdrawal"
+
+
+class TransactionManager():
```
I want you to find an other way for reconciliation of balance without changing previous transactions

Hamza-Chaudhary65 committed:
updated accounts model fields

Hamza-Chaudhary65 committed:
updates logic to reconcile account balance without updating previous transaction and creating new transaction
separate api for [get] description and [put, delete] transaction

sheralim012 commented on fintrack/accounts/management/commands/populate_data.py:
```diff
@@ -0,0 +1,35 @@
+            holder=user
+        )
+
+        trans1 = create_transaction("first transaction",datetime.date(2023,10,5),500, account, "Withdraw")
+
+        trans2 = create_transaction("secondst transaction",datetime.date(2023,10,5),5000, account, "Deposit")
```
I'm not sure if I'm able to communicate this in the right way but it's not good to keep on repeating mistakes again and again
read pep8 and format the code accordingly

sheralim012 commented on fintrack/accounts/models.py:
```diff
@@ -0,0 +1,23 @@
+
+
+class Account(models.Model):
+   
+
```
why these empty line?

sheralim012 commented on fintrack/choices.py:
```diff
@@ -0,0 +1,11 @@
+
+
+class TransactionCategoriesChoices(models.TextChoices):
+    Withdraw = "Withdraw", ("Withdraw")
```
even after putting the comment with the exact changes, still changes aren't made

sheralim012 commented on fintrack/transactions/templates/transactions/list_transactions.html:
are we still doing templates? why this that? aren't we making APIs?

sheralim012 commented on fintrack/transactions/views.py:
```diff
@@ -0,0 +1,61 @@
+@api_view(['GET'])
+def transaction_detail(request, pk):
+    try:
+        transaction = Transaction.objects.get(pk=pk)
```
read about get_object_or_404

sheralim012 commented on fintrack/transactions/views.py:
```diff
@@ -0,0 +1,61 @@
+        return Response (serializer.data)
+
+
+@api_view([ 'PUT', 'DELETE'])
```
as discussed yesterday, we wouldn't be changing existing transactions and always be doing more to handle revertion and updating case so why were are handling put and delete HTTP methods?

Hamza-Chaudhary65 committed:
implements updated api and handeler function for transaction
Transaction id and amount(optional) would be passed to revert a transaction.
if amount is also passed a seprate transaction for that amount would also take place in addition to revert trasaction

Hamza-Chaudhary65 committed:
deleted templates

Hamza-Chaudhary65 committed:
Uppercase

sheralim012 commented on fintrack/choices.py:
why these choices are in project and not the app?

sheralim012 commented on fintrack/fintrack/settings.py:
```diff
@@ -0,0 +1,134 @@
+
+from dotenv import load_dotenv
+
+load_dotenv()
```
use decouple package

Hamza-Chaudhary65 committed:
creates seprate choices file for each app.
uses decouple package

Hamza-Chaudhary65 committed:
Merge branch 'develop' into create_fintrack_views
